In [37]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

from ast import literal_eval
from itertools import islice

In [2]:
path = 'C:/Users/82109/Desktop/deep daiv. 스터디/2주차'

In [211]:
베스트셀러 = pd.read_excel(path + "/데이터 전처리/베스트셀러_최종2.xlsx")

In [212]:
best = 베스트셀러.iloc[:, 1:]
best

상품명  \
0                            위기의 역사   
1                돈의 속성(300쇄 리커버에디션)   
2                            부자의 그릇   
3                             사장학개론   
4                         EBS 다큐프라임   
...                             ...   
3708                         독서의 역사   
3709                             표준   
3710          삼국지연의보다 재미있는 정사 삼국지 1   
3711  나도 손글씨 바르게 쓰면 소원이 없겠네(핸디 워크북)   
3712                         성격의 탄생   

                                                    책소개  \
0     40만 독자가 선택한 오건영의 글로벌 경제사 특강\n누적 조회 수 550만 회 달성...   
1     2020ㆍ2021ㆍ2022ㆍ2023 4년 연속 최장기 베스트셀러\n80만 깨어있는 ...   
2     수많은 젊은 부자가 꼽은 최고의 ‘부자학 입문서’ 『부자의 그릇』 양장 개정판\n“...   
3     이 책은 과거와 현재,\n미래의 모든 사장을 위한 실무 지침서다\n이 책은 한국과 ...   
4     우리가 반드시 알아야 할 자본주의의 진실!\n『자본주의』는 자본주의를 쉽게 풀어낸 ...   
...                                                 ...   
3708  “독서를 다룬 책 중 가장 빼어난 이야기”\n언어의 파수꾼이자 책의 수호자, 세계 ...   
3709  제4판 표준 중세 국어 문법론에서는 지난 10년 동안에 나온 중세어 관련 문법 업적...   
3710  20만 구독자가 열광한 화제의 유튜브 역사채널 〈써에이스쇼〉의 삼국지\n누적 조회 ...   
3711  아무때나 시간 날 때 틈틈이\n지하철ㆍ학교ㆍ회사ㆍ카페ㆍ집 어느 장소에서도\n편하게 ...   
3712  내 성격은 어떻게 만들어진 걸까?\n왜 사람마다 성격이 다를까?\n성격을 바꿀 수 ...   

                                                    키워드  \
0     ['금리', '모기지', '흥국', '저금리', '글로벌 금융 위기', '실리콘밸리...   
1     ['부자되는법', '종잣돈', '회장', '투자 원칙', '경제철학', '주식 시장...   
2     ['부자되는법', '신용', '소설 형식', '배트', '경제이야기', '사업 아이...   
3         ['창업', '매출', '직원', '장학', '기업인', '사업', '경제경영']   
4     ['재테크', '경제학', '중앙은행', '케인스', '경제이론', '국부론', '...   
...                                                 ...   
3708  ['독서가', '움베르토 에코', '분서갱유', '지배 세력', '성서', '진시황...   
3709    ['품사론', '형태론', '통사론', '표기법', '겹문장', '발음', '인문']   
3710  ['판타지', '황건적', '유비', '컬러 삽화', '위촉', '실제 역사', '...   
3711     ['악필', '교정', '원고지', '정자체', '스프링', '지하철', '인문']   
3712     ['과학', '외향성', '유전학', '친화성', '규명', '개인차', '인문']   

                                                책소개 키워드  \
0     {'조회', '선택', '부활', '오건', '금융위기', '달성', '소멸', '...   
1     {'연속', '일본', '상황', '선택', '최', '속성', '장기', '추가'...   
2     {'채널', '로부터', '울림', '원리', '누구', '최고', '주먹밥', '...   
3     {'의', '저자', '정리', '존재', '운영', '모든', '장학', '구',...   
4     {'능력', '반드시', '앞', '필수', '모든', '진실', '세계', '활용...   
...                                                 ...   
3708  {'인간', '저자', '룸', '뒤', '분석', '자기', '형태', '위협',...   
3709  {'포함', '각주', '심화', '통해', '의미', '문법', '장', '표준'...   
3710  {'채널', '역사', '의', '저자', '돌파', '층', '총', '결의', ...   
3711  {'크기', '노트', '글씨', '스프링', '문장', '원고지', '가득', '...   
3712  {'인간', '저자', '존재', '사랑', '서로', '성실', '종합', '가치...   

                                                책소개 전처리  \
0     ['40만 독자가 선택한 오건영의 글로벌 경제사 특강', '누적 조회 수 550만 ...   
1     ['2020 2021 2022 2023 4년 연속 최장기 베스트셀러', '80만 깨...   
2     ['수많은 젊은 부자가 꼽은 최고의 부자학 입문서 부자의 그릇 양장 개정판', '당...   
3     ['이 책은 과거와 현재', '미래의 모든 사장을 위한 실무 지침서다', '이 책은...   
4     ['우리가 반드시 알아야 할 자본주의의 진실', '자본주의 는 자본주의를 쉽게 풀어...   
...                                                 ...   
3708  ['독서를 다룬 책 중 가장 빼어난 이야기', '언어의 파수꾼이자 책의 수호자 세계...   
3709  ['제4판 표준 중세 국어 문법론에서는 지난 10년 동안에 나온 중세어 관련 문법 ...   
3710  ['20만 구독자가 열광한 화제의 유튜브 역사채널 써에이스쇼 의 삼국지', '누적 ...   
3711  ['아무때나 시간 날 때 틈틈이', '지하철 학교 회사 카페 집 어느 장소에서도',...   
3712  ['내 성격은 어떻게 만들어진 걸까', '왜 사람마다 성격이 다를까', '성격을 바...   

                                            책소개 키워드 수정본  
0     ['부활', '인플레이션', '금융위기', '책', '경제위기', '소멸', '선택...  
1     ['기념', '돈', '속성', '바뀐', '맞춰', '20', '경제', '경영'...  
2     ['돈', '부자', '통해', '흥미로운', '의장', '스', '테디', '셀러...  
3     ['사장', '대한', '있는', '위', '가장', '되었다', '직원', '따라...  
4     ['자본주의', '있는', '그리고', '우리', '금융', '것', '오늘날', ...  
...                                                 ...  
3708  ['책', '있다', '저자', '대해', '관련', '된', '책', '독서', ...  
3709  ['포함', '문법', '제', '3', '편', '국어', '지난', '다루었다'...  
3710  ['삼국지', '정사', '역사', '시작', '장의', '마니아', '층', '있...  
3711  ['물론', '연습', '4', '주', '동안', '나도', '꾸준히', '종이'...  
3712  ['성격', '사람', '저자', '어떻게', '무엇', '일까', '뇌', '과학...  

[3713 rows x 6 columns]

In [213]:
best.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3713 entries, 0 to 3712
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   상품명          3713 non-null   object
 1   책소개          3713 non-null   object
 2   키워드          3713 non-null   object
 3   책소개 키워드      3713 non-null   object
 4   책소개 전처리      3713 non-null   object
 5   책소개 키워드 수정본  3713 non-null   object
dtypes: object(6)
memory usage: 174.2+ KB


In [214]:
best.isnull().sum()

상품명            0
책소개            0
키워드            0
책소개 키워드        0
책소개 전처리        0
책소개 키워드 수정본    0
dtype: int64

In [215]:
best.duplicated().sum()

0

#### 추천시스템 구현 과정

step 1. 도서를 키워드로 벡터화 + document-term matrix 생성

step 2. document-term matrix의 코사인 유사도 계산

step 3. 코사인 유사도 기반 추천리스트 작성

## 인코딩 기반 추천시스템


- 각 책에 대하여 키워드의 존재여부를 인코딩한다.
- 키워드 간 유사성 분석을 통해 동일한 키워드를 많이 가지고 있는 책을 우선적으로 추천하는 방식이다.

### CountVectorizer : CountVectorizer()로 document-term matrix 생성

In [40]:
from sklearn.feature_extraction.text import CountVectorizer

string 상태 그대로 적용

In [225]:
print(type(best.loc[0, '책소개 키워드 수정본']))

<class 'str'>


### 도서 벡터화

: document-term matrix 생성

### ver 0. '책소개 키워드 수정본' 사용

In [227]:
count_vectorizer = CountVectorizer()
book_mat_keyword = count_vectorizer.fit_transform(best['책소개 키워드 수정본'])   

In [228]:
book_mat_keyword.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

* dictionary 일부만 출력하기

```python
from itertools import islice
dict(islice(출력할 dictionary.items(), 출력 개수))
```

In [229]:
dict(islice(count_vect_keyword.vocabulary_.items(), 10))

{'금리': 952,
 '모기지': 2322,
 '흥국': 8748,
 '저금리': 6267,
 '글로벌': 940,
 '금융': 961,
 '위기': 5511,
 '실리콘밸리': 4456,
 '경제경영': 380,
 '부자되는법': 3133}

### ver 1. '키워드' 사용

In [234]:
count_vectorizer_1 = CountVectorizer()
book_mat_keyword_1 = count_vectorizer_1.fit_transform(best['키워드'])  

In [235]:
book_mat_keyword_1.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### ver 2. '상품명' 사용

In [245]:
count_vectorizer_2 = CountVectorizer()
book_mat_keyword_2 = count_vectorizer_2.fit_transform(best['상품명']) 

In [246]:
book_mat_keyword_2.toarray()

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### 코사인 유사도 계산

: document-term matrix의 각 행 사이 코사인 유사도 계산

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

In [247]:
cosine_sim = cosine_similarity(book_mat_keyword.toarray())
cosine_sim_1 = cosine_similarity(book_mat_keyword_1.toarray())
cosine_sim_2 = cosine_similarity(book_mat_keyword_2.toarray())

In [248]:
print(cosine_sim.shape)
print(cosine_sim[:10])

(3713, 3713)
[[1.         0.08944272 0.         ... 0.04152274 0.         0.        ]
 [0.08944272 1.         0.03849002 ... 0.03713907 0.         0.        ]
 [0.         0.03849002 1.         ... 0.10721125 0.03513642 0.03253   ]
 ...
 [0.         0.03592106 0.03456506 ... 0.         0.03279129 0.        ]
 [0.         0.03651484 0.03513642 ... 0.03390318 0.03333333 0.03086067]
 [0.         0.         0.         ... 0.03335187 0.03279129 0.03035884]]


### 유사도 가중치 설정

코사인 유사도 변수
> cosine_sim -> 책소개 키워드  
> cosine_sim_1 -> 키워드 Pick  
> cosine_sim_2 -> 상품명  

In [249]:
sim = 0.4*cosine_sim + 0.5*cosine_sim_1 + 0.1*cosine_sim_2
sim

array([[1.        , 0.09133264, 0.05555556, ..., 0.0166091 , 0.        ,
        0.        ],
       [0.09133264, 1.        , 0.12650712, ..., 0.01485563, 0.        ,
        0.        ],
       [0.05555556, 0.12650712, 1.        , ..., 0.0428845 , 0.01405457,
        0.013012  ],
       ...,
       [0.0166091 , 0.01485563, 0.0428845 , ..., 1.        , 0.06299408,
        0.07554938],
       [0.        , 0.        , 0.01405457, ..., 0.06299408, 1.        ,
        0.07142857],
       [0.        , 0.        , 0.013012  , ..., 0.07554938, 0.07142857,
        1.        ]])

## 추천리스트 작성

: 상위 top_n개의 유사 도서를 추천해는 함수 정의

추천 시스템 원리 = 주어진 title에 대하여 코사인 유사도가 높은 순부터 내림차순 정렬한 후 상위 n개를 추출하는 방식

**np.argsort()**
- 오름차순으로 정렬하는데, 이때 출력값으로 정렬되기 전의 인덱스 값을 리스트로 반환한다.  
- 역순 정렬을 통해 0~3712번 각 책과 유사한 책의 index를 내림차순으로 반환하도록 만든다.  
- 예를 들어 0번째 책의 경우 845번, 244번,,,, 1558번, 3712번 index 순서대로 유사도가 높다.

In [254]:
# 궁금한 도서의 index 입력
query_index = 0

In [263]:
def find_sim_book(best, sim, query_index, top_k):
    # query 인덱스에 대한 유사도 값을 가져옴
    query_similarity = sim[query_index]

    # 코사인 유사도 값을 기준으로 유사도가 높은 순서대로 정렬된 인덱스를 얻음
    sorted_indices = np.argsort(query_similarity)[::-1]

    print(f'검색한 콘텐츠 : {best["상품명"].iloc[query_index]}')
    print('\n출력한 콘텐츠')
    print(sorted_indices)
    print(best['상품명'].iloc[sorted_indices[1:top_k +1]])

In [264]:
find_sim_book(best, sim, query_index, top_k)

검색한 콘텐츠 : 위기의 역사

출력한 콘텐츠
[   0  845  244 ... 1560 1558 3712]
845              부채의 역습
244     세계화의 종말과 새로운 시작
880    글로벌경제 상식사전(2023)
487       경제 상식사전(2022)
806              투자의 모험
400              좋은 불평등
96            투자에 대한 생각
459           EBS 다큐프라임
177             맨큐의 경제학
108      돈을 찍어내는 제왕, 연준
Name: 상품명, dtype: object


CountVectorizer의 문제점
- CountVectorizer()를 실행하면서 이미 완료된 키워드 토큰화가 다시 토큰화되는 문제  
- 아래 예시의 경우 0행 키워드로 '글로벌 금융 위기'가 존재하지만 `count_vect_keyword.vocabulary_`에는 존재하지 않음

In [210]:
best['키워드'][0]
#count_vect_keyword.vocabulary_['글로벌 금융 위기']

"['금리', '모기지', '흥국', '저금리', '글로벌 금융 위기', '실리콘밸리', '경제경영']"

### 원 핫 인코딩 : 원 핫 인코딩 진행해 document-term matrix 생성

string이 아닌 list로 인식되도록 변경

In [271]:
print(type(best.loc[0, '키워드']))

best2=best.copy()
best2['키워드'] = best2['키워드'].apply(lambda x: literal_eval(x))
best2['책소개 키워드 수정본'] = best2['책소개 키워드 수정본'].apply(lambda x: literal_eval(x))

print(type(best2.loc[0, '키워드']))
print(type(best2.loc[0, '책소개 키워드 수정본']))

<class 'str'>
<class 'list'>
<class 'list'>


### 도서 벡터화

: document-term matrix 생성

step 1. 키워드 dictionary 만들기 : 존재하는 모든 키워드 list `word_dic` 생성

In [272]:
best2.head()

상품명                                                책소개  \
0              위기의 역사  40만 독자가 선택한 오건영의 글로벌 경제사 특강\n누적 조회 수 550만 회 달성...   
1  돈의 속성(300쇄 리커버에디션)  2020ㆍ2021ㆍ2022ㆍ2023 4년 연속 최장기 베스트셀러\n80만 깨어있는 ...   
2              부자의 그릇  수많은 젊은 부자가 꼽은 최고의 ‘부자학 입문서’ 『부자의 그릇』 양장 개정판\n“...   
3               사장학개론  이 책은 과거와 현재,\n미래의 모든 사장을 위한 실무 지침서다\n이 책은 한국과 ...   
4           EBS 다큐프라임  우리가 반드시 알아야 할 자본주의의 진실!\n『자본주의』는 자본주의를 쉽게 풀어낸 ...   

                                           키워드  \
0   [금리, 모기지, 흥국, 저금리, 글로벌 금융 위기, 실리콘밸리, 경제경영]   
1   [부자되는법, 종잣돈, 회장, 투자 원칙, 경제철학, 주식 시장, 경제경영]   
2  [부자되는법, 신용, 소설 형식, 배트, 경제이야기, 사업 아이템, 경제경영]   
3              [창업, 매출, 직원, 장학, 기업인, 사업, 경제경영]   
4       [재테크, 경제학, 중앙은행, 케인스, 경제이론, 국부론, 경제경영]   

                                             책소개 키워드  \
0  {'조회', '선택', '부활', '오건', '금융위기', '달성', '소멸', '...   
1  {'연속', '일본', '상황', '선택', '최', '속성', '장기', '추가'...   
2  {'채널', '로부터', '울림', '원리', '누구', '최고', '주먹밥', '...   
3  {'의', '저자', '정리', '존재', '운영', '모든', '장학', '구',...   
4  {'능력', '반드시', '앞', '필수', '모든', '진실', '세계', '활용...   

                                             책소개 전처리  \
0  ['40만 독자가 선택한 오건영의 글로벌 경제사 특강', '누적 조회 수 550만 ...   
1  ['2020 2021 2022 2023 4년 연속 최장기 베스트셀러', '80만 깨...   
2  ['수많은 젊은 부자가 꼽은 최고의 부자학 입문서 부자의 그릇 양장 개정판', '당...   
3  ['이 책은 과거와 현재', '미래의 모든 사장을 위한 실무 지침서다', '이 책은...   
4  ['우리가 반드시 알아야 할 자본주의의 진실', '자본주의 는 자본주의를 쉽게 풀어...   

                                         책소개 키워드 수정본  \
0  [부활, 인플레이션, 금융위기, 책, 경제위기, 소멸, 선택, 생, 성과, 오건, ...   
1  [기념, 돈, 속성, 바뀐, 맞춰, 20, 경제, 경영, 필, 도서, 추가, 최, ...   
2  [돈, 부자, 통해, 흥미로운, 의장, 스, 테디, 셀러, 돈, 돈, 젊은, 가게,...   
3  [사장, 대한, 있는, 위, 가장, 되었다, 직원, 따라, 지금, 하는지, 한정, ...   
4  [자본주의, 있는, 그리고, 우리, 금융, 것, 오늘날, 쉽게, 진실, 것, 인지도...   

                                  상품명_전처리  
0                            ['위기', '역사']  
1  ['돈', '속성', '쇄', '리', '커버', '에디', '션']  
2                            ['부자', '그릇']  
3                            ['장학', '개론']  
4                               ['다큐프라임']

In [273]:
word=list()

for i in range(len(best2['키워드'])):
    keyword_list = best2.loc[i, '키워드']
    word += keyword_list

word_dic = list(set(word))

In [274]:
len(word_dic)

10465

In [275]:
type(word_dic)

list

step 2. 벡터화 함수 설정 : 각 도서를 키워드로 빈도수로 벡터화하는 함수 정의

In [173]:
tmp = []
lst = ['금리', '모기지', '흥국', '저금리', '글로벌 금융 위기', '실리콘밸리', '경제경영']

for voca in word_dic:
        tmp.append(lst.count(voca))

In [174]:
len([i for i in tmp if i>0])

7

list.count() method를 이용해 키워드의 존재여부에 대한 DTM dataframe 생성

In [182]:
def vectorize(df, col):
    df_matrix = pd.DataFrame()
    #columns=range(10465)
    #columns = list(word_dic)

    for i in range(df.shape[0]):
        tmp = []
        for voca in word_dic:
            tmp.append(df.loc[i, col].count(voca))
        df_matrix = pd.concat([df_matrix,  pd.DataFrame(tmp).T])
        #df_matrix.append(tmp)
    return df_matrix

### ver 0. '책소개 키워드 수정본' 사용

In [276]:
df_matrix = vectorize(best2, '책소개 키워드 수정본')

In [277]:
df_matrix.reset_index(drop=True, inplace=True)

In [278]:
df_matrix

0      1      2      3      4      5      6      7      8      9      \
0         0      0      0      0      0      0      0      0      0      0   
1         0      0      0      0      0      0      0      0      0      0   
2         0      0      0      0      0      0      0      0      0      0   
3         0      0      0      0      0      0      0      0      0      0   
4         0      0      0      0      0      0      0      0      0      0   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708      0      0      0      0      0      0      0      0      0      0   
3709      0      0      0      0      0      0      0      0      0      0   
3710      0      0      0      0      0      0      0      0      0      0   
3711      0      0      0      0      0      0      0      0      0      0   
3712      0      0      0      0      0      0      0      0      0      0   

      ...  10455  10456  10457  10458  10459  10460  10461  10462  10463  \
0     ...      0      0      0      0      0      0      0      0      0   
1     ...      0      0      0      0      0      0      0      0      0   
2     ...      0      0      0      0      0      0      0      0      0   
3     ...      0      0      0      0      0      1      0      0      0   
4     ...      0      0      0      0      0      0      0      0      0   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708  ...      0      0      0      0      0      0      0      0      0   
3709  ...      0      0      0      0      0      0      0      0      0   
3710  ...      0      0      0      0      0      0      0      0      0   
3711  ...      0      0      0      0      0      0      0      0      0   
3712  ...      0      0      0      0      0      0      0      0      0   

      10464  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
3708      0  
3709      0  
3710      0  
3711      0  
3712      0  

[3713 rows x 10465 columns]

In [279]:
# 해당되는 키워드에 대해 모두 1의 값이 들어갔는지 확인
for i in best2.loc[2, '키워드']:
    print(df_matrix_c.iloc[2, list(word_dic).index(i)])

1
1
1
1
1
1
1


dataframe을 DTM matrix로 변환

In [280]:
book_mat = df_matrix.to_numpy()
book_mat

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### ver 1. '키워드' 사용

In [283]:
df_matrix_1 = vectorize(best2, '키워드')

In [284]:
df_matrix_1.reset_index(drop=True, inplace=True)
df_matrix_1

0      1      2      3      4      5      6      7      8      9      \
0         0      0      0      0      0      0      0      0      0      0   
1         0      0      0      0      0      0      0      0      0      0   
2         0      0      0      0      0      0      0      0      0      0   
3         0      0      0      0      0      0      0      0      0      0   
4         0      0      0      0      0      0      0      0      0      0   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708      0      0      0      0      0      0      0      0      0      0   
3709      0      0      0      0      0      0      0      0      0      0   
3710      0      0      0      0      0      0      0      0      0      0   
3711      0      0      0      0      0      0      0      0      0      0   
3712      0      0      0      0      0      0      0      0      0      0   

      ...  10455  10456  10457  10458  10459  10460  10461  10462  10463  \
0     ...      0      0      0      0      0      0      0      0      0   
1     ...      0      0      0      0      0      0      0      0      0   
2     ...      0      0      0      0      0      0      0      0      0   
3     ...      0      0      0      0      0      1      0      0      0   
4     ...      0      0      0      0      0      0      0      0      0   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708  ...      0      0      0      0      0      0      0      0      0   
3709  ...      0      0      0      0      0      0      0      0      0   
3710  ...      0      0      0      0      0      0      0      0      0   
3711  ...      0      0      0      0      0      0      0      0      0   
3712  ...      0      0      0      0      0      0      0      0      0   

      10464  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
3708      0  
3709      0  
3710      0  
3711      0  
3712      0  

[3713 rows x 10465 columns]

In [285]:
book_mat_1 = df_matrix_1.to_numpy()
book_mat_1

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### ver 1. '상품명' 사용

In [286]:
df_matrix_2 = vectorize(best2, '상품명')

In [287]:
df_matrix_2.reset_index(drop=True, inplace=True)
df_matrix_2

0      1      2      3      4      5      6      7      8      9      \
0         0      0      0      0      0      0      0      0      0      0   
1         0      0      0      0      0      0      0      0      0      0   
2         0      0      0      0      0      0      0      0      0      0   
3         0      0      0      0      0      0      0      0      0      0   
4         0      0      0      0      0      0      0      0      0      0   
...     ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708      0      0      0      0      0      0      0      0      0      0   
3709      0      0      0      0      0      0      0      0      0      0   
3710      0      0      0      0      0      0      0      0      0      0   
3711      0      0      0      0      0      0      0      0      0      0   
3712      0      0      0      0      0      0      0      0      0      0   

      ...  10455  10456  10457  10458  10459  10460  10461  10462  10463  \
0     ...      0      0      0      0      0      0      0      0      0   
1     ...      0      0      0      0      0      0      0      0      0   
2     ...      0      0      0      0      0      0      0      0      0   
3     ...      0      0      0      0      0      0      0      0      0   
4     ...      0      0      0      0      0      0      0      0      0   
...   ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
3708  ...      0      0      0      0      0      0      0      0      0   
3709  ...      0      0      0      0      0      0      0      0      0   
3710  ...      0      0      0      0      0      0      0      0      0   
3711  ...      0      0      0      0      0      0      0      0      0   
3712  ...      0      0      0      0      0      0      0      0      0   

      10464  
0         0  
1         0  
2         0  
3         0  
4         0  
...     ...  
3708      0  
3709      0  
3710      0  
3711      0  
3712      0  

[3713 rows x 10465 columns]

In [288]:
book_mat_2 = df_matrix_2.to_numpy()
book_mat_2

array([[0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       ...,
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0],
       [0, 0, 0, ..., 0, 0, 0]], dtype=int64)

### 코사인 유사도 계산

: document-term matrix의 각 행 사이 코사인 유사도 계산

In [289]:
cosine_sim_re = cosine_similarity(book_mat)
cosine_sim_1_re = cosine_similarity(book_mat_1)
cosine_sim_2_re = cosine_similarity(book_mat_2)

In [290]:
print(cosine_sim_re.shape)
print(cosine_sim_re[:10])

(3713, 3713)
[[1.         0.08451543 0.         ... 0.         0.         0.        ]
 [0.08451543 1.         0.3678836  ... 0.         0.         0.        ]
 [0.         0.3678836  1.         ... 0.         0.         0.        ]
 ...
 [0.         0.0805823  0.         ... 0.         0.         0.        ]
 [0.         0.0805823  0.06917145 ... 0.         0.         0.        ]
 [0.         0.         0.         ... 0.         0.         0.070014  ]]


### 유사도 가중치 설정

코사인 유사도 변수
> cosine_sim_re -> 책소개 키워드  
> cosine_sim_1_re -> 키워드 Pick  
> cosine_sim_2_re -> 상품명  

In [291]:
sim_re = 0.4*cosine_sim_re + 0.5*cosine_sim_1_re + 0.1*cosine_sim_2_re
sim_re

array([[1.        , 0.10523474, 0.07142857, ..., 0.        , 0.        ,
        0.        ],
       [0.10523474, 1.        , 0.29001058, ..., 0.        , 0.        ,
        0.        ],
       [0.07142857, 0.29001058, 1.        , ..., 0.        , 0.        ,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.07142857,
        0.07142857],
       [0.        , 0.        , 0.        , ..., 0.07142857, 1.        ,
        0.07142857],
       [0.        , 0.        , 0.        , ..., 0.07142857, 0.07142857,
        1.        ]])

## 추천리스트 작성

: 상위 top_n개의 유사 도서를 추천해는 함수 정의

In [292]:
# 궁금한 도서의 index, 추천 받을 책 개수 입력
query_index = 0
top_k = 10

In [293]:
find_sim_book(best, sim_re, query_index, top_k)

검색한 콘텐츠 : 위기의 역사

출력한 콘텐츠
[   0  845  244 ... 2422 2421 1856]
845                부채의 역습
244       세계화의 종말과 새로운 시작
487         경제 상식사전(2022)
107                초거대 위협
757    난처한 경제 이야기 3: 금융 편
806                투자의 모험
553              다이버시티 파워
298           미국 자본주의의 역사
150           경제지표를 읽는 시간
108        돈을 찍어내는 제왕, 연준
Name: 상품명, dtype: object


## 생성된 DTM과 코사인 유사도 matrix 총정리

방법 1. CountVectorizer 이용한 벡터화

> ver 1. cosine_sim / book_mat_keyword  
> ver 2. cosine_sim_1 / book_mat_keyword_1  
> ver 3. cosine_sim_2 / book_mat_keyword_2


방법 2. 원-핫 인코딩을 이용한 벡터화

> ver 1. cosine_sim_re / book_mat
> ver 2. cosine_sim_1_re / book_mat_1  
> ver 3. cosine_sim_2_re / book_mat_2

### 방법 및 버전에 따른 추천 결과 비교

In [299]:
query_index = 10
top_k = 10

In [300]:
find_sim_book(best, sim, query_index, top_k)

검색한 콘텐츠 : 어떤 하락장에서도 살아남을 수 있는

출력한 콘텐츠
[  10  650  839 ... 2285 2284 3712]
650            세력주 매매 공식
839         이기는 투자의 심리법칙
803     주식 대세판단 무작정 따라하기
368    뉴욕주민의 진짜 미국식 주식투자
867      일본 주식시장의 승부사들 1
89           여의도 선수들의 비밀
245          최강의 배당연금 투자
481               투자의 원칙
141        피터 린치의 투자 이야기
563                   세력
Name: 상품명, dtype: object


In [301]:
find_sim_book(best, sim_re, query_index, top_k)

검색한 콘텐츠 : 어떤 하락장에서도 살아남을 수 있는

출력한 콘텐츠
[  10  650  803 ... 1716 1715 3712]
650            세력주 매매 공식
803     주식 대세판단 무작정 따라하기
368    뉴욕주민의 진짜 미국식 주식투자
89           여의도 선수들의 비밀
481               투자의 원칙
839         이기는 투자의 심리법칙
53          세력주의 신 100법칙
867      일본 주식시장의 승부사들 1
245          최강의 배당연금 투자
319        살 때, 팔 때, 벌 때
Name: 상품명, dtype: object


In [314]:
query_index = 2370
top_k = 10
find_sim_book(best, sim_re, query_index, top_k)

검색한 콘텐츠 : 어린 왕자(Le Petit Prince)(교보문고 특별판)

출력한 콘텐츠
[2370 1975 2775 ... 1821 1820    0]
1975                              어린 왕자
2775                        셜록 홈즈 전집 세트
1977                            노르웨이의 숲
2284                코드612 누가 어린왕자를 죽였는가
2684                              애린 왕자
2251                                금각사
2723                          어린왕자(미니북)
2363                                이방인
2045                        그리고 아무도 없었다
2233    싯다르타(1922년 오리지널 초판본 표지디자인)(문고판)
Name: 상품명, dtype: object
